[feat] NFT.storage continue...

melMass opened the pull request:
Fixing issues due to the IPFS switch and older unreported ones:

- [x] Cannot edit profile
   - [x] Works but there is absolutely no feedback...
- [x] MD preview / styling -> fixed displayURi issue, styling should be done is a dedicated PR
- [x] Subfolders for generative tokens (linked to https://github.com/teia-community/ipfs-upload-proxy/pull/3)
- [x] ~~Clean the folders before upload? (macOS creates .DS_Store files and `._.` prefixed files, windows has ThumbDB), might be "dangerous" to alter the archive~~

melMass committed:
fix: ⚡️ use MIMETYPES everywhere

melMass committed:
fix: ⚡️ jsconfig

melMass committed:
fix: ⚡️ move function to utils

melMass committed:
fix: ⚡️ wip remove all references to Infura

melMass committed:
fix: 🎨 prevent uploading broken meta

melMass committed:
fix: ⚡️ further removing of infura instances

melMass committed:
fix: ⚡️ wip use HashToURL everywhere

melMass commented:
I was about to remove every instance of IPFS as we use various gateways.
@NoRulesJustFeels while doing that I wonder if this won't make reaching gateways limits faster?

For now I just changed the SUBJKT icon (was DWEB -> IPFS.io), keeping the current gateway used by others, we can switch easily once we decide

melMass committed:
fix: 🎨 build error

Also add NFTStorage as suggested by NRJF in #150.

melMass committed:
fix: ⚡️ use CIDtoURL

LeonNicholls commented:
As part of the discussion on #150, there are various IPFS options to consider. We should probably spend some time testing those when we have access. We might even consider some kind of failover solution for the client if the rate limit is reached for a particular gateway.

melMass commented:
> As part of the discussion on #150, there are various IPFS options to consider. We should probably spend some time testing those when we have access. We might even consider some kind of failover solution for the client if the rate limit is reached for a particular gateway.

I was not up to date with that issue! Awesome

melMass committed:
fix: ⚡️ imports

melMass committed:
fix: ⚡️ wrong and not used

This was setting a .gravatar property to the parsed metadatafile.
Does not make sense and is not used anywhere.

melMass committed:
fix: ⚡️ use an env variable to set the IPFS gateway

The optional second param for HashToURL and CIDToURL
allows you to override the IPFS gateway to use.

melMass committed:
fix: ⚡️ subjkt feedback

melMass committed:
fix: ⚡️ context

- use const where it makes sense (easier to read)
- fixed the getBalance function even if it should not be there...
- added a fetchJSON util function.

melMass committed:
feat: 🔥 warn if balance is low on mint

melMass commented:
## Added a few things:

- **Check the balance on the mint page**: This displays if you have less than 0.15Tz (6Tz for the screenshot)
	<img width="400" alt="image" src="https://user-images.githubusercontent.com/7041726/184462975-214ef0b6-9c35-4e29-8d6c-534cc49b488e.png">
	
- **Check for existing SUBJKT**: This is an ongoing issue people had as there is no check before making the contract call so it only failed in Temple/Kukai not before. Now it's showing the owner of the name:
	<img width="400" alt="image" src="https://user-images.githubusercontent.com/7041726/184463073-fe0c9541-2803-4348-91cc-2525ac505d20.png">

- **Replaced alert() with the context feedback**:
	<img width="400" alt="image" src="https://user-images.githubusercontent.com/7041726/184499200-06cb5885-9a03-44ed-8f7c-60e50b3a5642.png">


- All IPFS **CIDs** and **Hashes** now use dedicated util functions shared across to convert them to http gateway links. The second argument to specify the gateway is optional and by default uses a new dedicated ENV Variable (`REACT_APP_IPFS_DEFAULT_GATEWAY`). A few different gateways (DWeb, Cloudflare, Ipfs.io, Infura) where used before.

melMass committed:
fix: ⚡️ fetch more token in the batch

Seems to fix #110, but still unexplained...

melMass committed:
fix: 🎨 add (optional) to optional fields

melMass committed:
fix: ⚡️ use showfeedback, fix fields

melMass committed:
fix: ✨ reformat debug logs to take less space

melMass committed:
fix: ⚡️ avoid the warning glitch on first mount

melMass committed:
fix: ⚡️ use showFeedback

melMass committed:
fix: ⚡️ prepare for v6 with fallback for v5

fixes #160

melMass committed:
feat: ⚡️ add long form description

melMass commented:
Ah ok... The CI is not building to the stage URL since quite a few commits, related to this somehow even if we use npm everywhere
https://github.com/dilanx/craco/issues/435

melMass committed:
fix: ⚡️ CI Builds

Adding typescript as a dev dependency.
This is a bug upstream in craco:
https://github.com/dilanx/craco/issues/435

melMass committed:
test: ⚡️ use nft.storage gateway (temp)

melMass committed:
fix: 🎨 use hash instead of cid method

xat commented on jsconfig.json:
```diff
@@ -1,14 +1,18 @@
 {
```
out of interest: why the changes here?

xat commented on src/components/collab/show/CollabDisplay.js:
```diff
@@ -88,13 +89,9 @@ export const CollabDisplay = () => {
-          .then(({ data }) => {
-            setLogo(data.identicon)
-          })
+        axios.get(HashToURL(metadataFile)).then(({ data }) => {
```
i think it would be better to use `await` here, otherwise errors won't be propagated to the `catch()` function that comes after `fetchMeta()`

xat commented on src/components/media-types/html/index.js:
```diff
@@ -54,24 +54,31 @@ export const HTMLComponent = (props) => {
+      const files = {}
+      filesArr.forEach((f) => {
+        files[f.path] = f.blob
+      })
```
not important (it's also not new code), but Array.reduce comes in handy in these cases:

```JavaScript
const files = filesArr.reduce((memo, f) => ({ ...memo, [f.path]: f.blob }), {});
```

xat commented on src/components/media-types/index.js:
```diff
@@ -93,8 +59,8 @@ export const renderMediaType = ({
 }) => {
-  let parsedArtifactUri
-  let parsedDisplayUri
+  const parsedArtifactUri = artifactUri ? HashToURL(artifactUri) : ''
```
i think the `HashToURL` function needs a second argument.

melMass replied:
By default it's optional and using the ENV key to choose the gateway, the second argument can be used to specify the gateway per instance use

xat replied:
ah, true, i missed that 👍

xat commented on src/components/media-types/video/index.js:
```diff
@@ -49,37 +47,33 @@ export const VideoComponent = ({
+      />
+    </div>
+  ) : (
+    <>
```
not important, but the `<></>` wrapping is not needed (i know, it was there before)

xat commented on src/constants.js:
```diff
@@ -90,32 +90,31 @@ export const MINT_FILESIZE = 100
-  WEBM: 'video/webm',
   GLB: 'model/gltf-binary',
   GLTF: 'model/gltf+json',
+  IPFS_DIRECTORY: 'application/x-directory',
```
maybe just `DIRECTORY`, to make it consitent with the other mimetype constants.

xat commented on src/pages/config/index.js:
```diff
@@ -113,29 +128,112 @@ export class Config extends Component {
       const buffer = Buffer.from(await file.arrayBuffer())
-      console.log(buffer)
-      this.setState({ identicon: 'ipfs://' + (await ipfs.add(buffer)).path })
+      const picture_cid = await uploadFileToIPFSProxy({
```
i'm not sure, but can't `file` just be passed directly to `uploadFileToIPFSProxy`?

melMass replied:
Probably!

This was a bit of the trouble in this PR, that's why I roughly typed it `@typedef { {path: string?, blob: Blob} } FileHolder`, extracting it from the old code and it's usage mainly to make sense of the code. I would happily clean it further in a separate branch/PR later

xat replied:
ok, yeah, no need that we need to block merging this pr because of that.

xat commented on src/pages/objkt-display/index.js:
```diff
@@ -125,6 +127,36 @@ export const ObjktDisplay = () => {
 
   const Tab = TABS[tabIndex].component
+
+  const objkt_classes = []
```
i think this needs to be put in `useState`.

melMass replied:
Isn't making the useEffect depend on nft alleviates that?

melMass replied:
I've done all besides this as I'm not sure ;)

xat replied:
hmm, well, if it works, it works :)

as soon as you add something async to the effect, it will probably break though.

xat reviewed (commented):
left a few comments, nothing dramatic :)

melMass commented:
> left a few comments, nothing dramatic :)

Thank youu, making the edits now!

melMass committed:
fix: ⚡️ xat feedbacks

xat commented:
i manually tested the "Cannot edit profile" fix, it worked 👍 

if there is nothing left otherwise, i can merge.

melMass commented:
> i manually tested the "Cannot edit profile" fix, it worked 👍
> 
> if there is nothing left otherwise, i can merge.

Yep it solves a few important things like subfolders for generative and profile editing. Further edits can be made in another one ;)

xat reviewed (approved):
great work!
